### PR TITLE
feat!(runtime): Introduce `pallet-gear-bank`

### PR DESCRIPTION
**Pull request by breathx**
Introducing a new pallet pallet-gear-bank and integrating it into the existing pallets logic.

### TODO:
- [x] Integrate `pallet-gear-bank` into pallets logic
- [x] Rebuild gsdk api
- Write migrations:
    - [x] Move all reserved funds to `pallet-gear-bank`
    - [x] Create bank account with ED balance

### CODE REVIEW

File: runtime/gear/src/lib.rs [hunk 1]
```diff
@@ -449,14 +449,30 @@ parameter_types! {
...
+fn bank_address() -> AccountId {
+    let arr: [u8; 8] = [1, 5, 0, 8, 2, 0, 0, 1];
+    let mut addr = [0u8; 32];
+
+    for (pos, byte) in arr.into_iter().enumerate() {
+        addr[pos * 4..pos * 4 + 4].copy_from_slice(&byte.to_le_bytes());
+    }
+
+    addr.into()
+}
```
Comment (gshep) [inline runtime/gear/src/lib.rs:461]: Why don't you use something like `b"gearge...argearbank"`/`b"gearbankgearba...nkgearbank"`?

Moreover I would place this common address in `runtime/common`.

**Review by gshep — APPROVED** [review]

File: runtime/gear/src/lib.rs (hunk 1 above)
Comment (ukint-vs) [reply runtime/gear/src/lib.rs:461]: @ekovalev advise here on entropy please. We need to be super secure

**Review by ukint-vs — APPROVED** [review]

**Review by ukint-vs — APPROVED** [review]
great

File: node/service/src/chain_spec/vara.rs [hunk 2]
```diff
@@ -521,6 +523,7 @@ pub fn main() -> Result<ChainSpec, String> {
                 vec![
                     // root_key
                     hex!["2455655ad2a1f9fbe510699026fc810a2b3cb91d432c141db54a9968da944955"].into(),
+                    // TODO: append here bank account
```
Comment (NikVolf) [inline node/service/src/chain_spec/vara.rs:526]: Better be type-safe and _require_ bank account id on caller

File: runtime/common/src/migrations.rs [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+    fn pre_upgrade() -> Result<Vec<u8>, &'static str> {
+        Ok(Default::default())
+    }
+
+    fn on_runtime_upgrade() -> Weight {
+        let version = T::Version::get().spec_version;
+
+        log::info!("🚚 Running migration to gear-bank with current spec version {version:?}");
+
+        // TODO: consider version here.
```
Comment (NikVolf) [inline runtime/common/src/migrations.rs:56]: when this TODO should be done?

**Review by ark0f — APPROVED** [review]

File: common/src/gas_provider/node.rs [hunk 4]
```diff
@@ -203,6 +203,21 @@ pub struct ChildrenRefs {
...
+    /// Returns total gas value inside GasNode.
+    pub fn total_value(&self) -> Balance {
+        self.value()
+            .unwrap_or_default()
+            .saturating_add(self.lock().total_locked())
+            .saturating_add(self.system_reserve().unwrap_or_default())
+    }
+}
+
 impl<ExternalId: Clone, Id: Clone + Copy, Balance: Default + Zero + Clone + Copy>
```
[the remark this answers is not in the document]
Comment (breathx) [reply common/src/gas_provider/node.rs:221]: No need in other impl Balance to be Saturating - it causes some other trait bounds problem

File: pallets/gear-bank/src/lib.rs [hunk 5]
```diff
@@ -0,0 +1,403 @@
...
+            }
+
+            Self::withdraw_value_no_transfer(account_id, value)?;
+
+            Self::withdraw(destination, value).unwrap_or_else(|e| unreachable!("qed above: {e:?}"));
+
+            Ok(())
+        }
+
+        pub fn account_gas(account_id: &AccountIdOf<T>) -> Option<BalanceOf<T>> {
```
[the remark this answers is not in the document]
Comment (breathx) [reply pallets/gear-bank/src/lib.rs:389]: Why?

File: runtime/common/src/migrations.rs [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+            // Depositing value.
+            let accounts_iter = AccountsOf::<T>::iter();
+            for (account_id, AccountInfo { data, .. }) in accounts_iter {
+                let reserve = data.reserved;
+                if !reserve.is_zero() {
+                    Balances::<T>::unreserve(&account_id, reserve);
+                }
+            }
+
+            T::DbWeight::get().writes(ops)
```
[the remark this answers is not in the document]
Comment (breathx) [reply runtime/common/src/migrations.rs:124]: Mentioned in approximation comment

File: pallets/gear-bank/src/lib.rs [hunk 7]
```diff
@@ -0,0 +1,403 @@
...
+    impl<Balance: Zero> Zero for BankAccount<Balance> {
+        fn zero() -> Self {
+            Self {
+                gas: Zero::zero(),
+                value: Zero::zero(),
+            }
+        }
+
+        fn is_zero(&self) -> bool {
+            self.gas.is_zero() && self.value.is_zero()
```
[the remark this answers is not in the document]
Comment (breathx) [reply pallets/gear-bank/src/lib.rs:138]: Whats the point
it requires to perform more operations other than current approach

File: runtime/common/src/migrations.rs (hunk 3 above)
Comment (breathx) [reply runtime/common/src/migrations.rs:56]: Once we decide when will we perform runtime upgrade containing this migrations

**Review by NikVolf — APPROVED** [review]

**Comment by shamilsan** [comment]
@CodiumAI-Agent /review

**Comment by shamilsan** [comment]
@CodiumAI-Agent /improve --extended

**Pull request merged**